lib/spack/spack/spec.py

```diff
@@ -769,13 +770,130 @@ def __str__(self):
             for key in sorted_keys) + cond_symbol
 
 
-class DependencyMap(lang.HashableMap):
-    """Each spec has a DependencyMap containing specs for its dependencies.
-       The DependencyMap is keyed by name. """
+def _sort_by_dep_types(dspec):
+    # Use negation since False < True for sorting
+    return (
+        'link' not in dspec.deptypes,
+        'run' not in dspec.deptypes,
+        'build' not in dspec.deptypes,
+        'test' not in dspec.deptypes
+    )
+
+
+@lang.lazy_lexicographic_ordering
+class _EdgeMap(Mapping):
+    """Represent a collection of edges (DependencySpec objects) in the DAG.
+
+    Objects of this class are used in Specs to track edges that are
+    outgoing towards direct dependencies, or edges that are ingoing
+    from direct dependents.
+
+    Edges are stored in a dictionary and keyed by package name.
+    """
+    def __init__(self, store_by='child'):
+        # Sanitize input arguments
+        msg = 'unexpected value for "store_by" argument'
+        assert store_by in ('child', 'parent'), msg
+
+        self.edges = {}
+        self.store_by_child = (store_by == 'child')
+
+    def __getitem__(self, key):
+        return self.edges[key]
+
+    def __iter__(self):
+        return iter(self.edges)
+
+    def __len__(self):
+        return len(self.edges)
+
+    def add(self, edge):
+        """Adds a new edge to this object.
+
+        Args:
+            edge (DependencySpec): edge to be added
+        """
+        key = edge.spec.name if self.store_by_child else edge.parent.name
+
+        current_list = self.edges.setdefault(key, [])
+
+        # The list of edges is empty, just append the edge
+        if not current_list:
+            current_list.append(edge)
+            return
+
+        # If there's something in the list, check if we need to update an
+        # already existing entry.
+        for dep in current_list:
+            if edge.spec == dep.spec and edge.parent == dep.parent:
+                dep.add_type(edge.deptypes)
+                break
+        else:
+            current_list.append(edge)
+
+        # Keep the list sorted
+        current_list.sort(key=_sort_by_dep_types)
 
     def __str__(self):
         return "{deps: %s}" % ', '.join(str(d) for d in sorted(self.values()))
 
+    def _cmp_iter(self):
+        for item in sorted(itertools.chain.from_iterable(self.edges.values())):
+            yield item
+
+    def copy(self):
+        """Copies this object and returns a clone"""
+        clone = type(self)()
+        clone.store_by_child = self.store_by_child
+
+        # Copy everything from this dict into it.
+        for dspec in itertools.chain.from_iterable(self.values()):
+            clone.add(dspec.copy())
+
+        return clone
+
+    def select_by(self, parent=None, child=None, dependency_types='all'):
```

Comment:
I think the API is correct, in the sense that an `_EdgeMap` can store arbitrary list of edges, which are collected in a dictionary by either parent name or child name. What I mean is that it's specs that use `_EdgeMap` objects to store edges having all the same parent or all the same children but an `_EdgeMap` object can in principle store completely different edges in terms of parents and children. Hence I think on this class the API is fine.

Reply:
> This API seems a little strange to me because the edge map is entirely parents or entirely children.

Think for example storing all the edges for an environment that has multiple root specs and wanting to retrieve all the different edges from e.g. `hdf5` to `zlib`. With the current API you can do it with this object.

Another way of saying this is that the current API fit using an EdgeMap object outside of a Spec object as a collection of edges.